File: src/rules/pylint/rules/invalid_all_format.rs
use ruff_macros::derive_message_formats;
use rustpython_ast::Expr;

use crate::ast::types::Range;
use crate::checkers::ast::Checker;
use crate::define_simple_violation;
use crate::registry::Diagnostic;
use crate::violation::Violation;

define_simple_violation!(
    InvalidAllFormat,
    "Invalid format for `__all__`, must be `tuple` or `list`"
);

/// PLE0605
pub fn invalid_all_format(checker: &mut Checker, expr: &Expr) {
    checker
        .diagnostics
        .push(Diagnostic::new(InvalidAllFormat, Range::from_located(expr)));
}

File: crates/ruff_linter/resources/test/fixtures/pylint/invalid_all_format.py
__all__ = "CONST"  # [invalid-all-format]

__all__ = ["Hello"] + {"world"}  # [invalid-all-format]

__all__ += {"world"}  # [invalid-all-format]

__all__ = {"world"} + ["Hello"]  # [invalid-all-format]

__all__ = {"world"} + list(["Hello"])  # [invalid-all-format]

__all__ = list(["Hello"]) + {"world"}  # [invalid-all-format]

__all__ = (x for x in ["Hello", "world"])  # [invalid-all-format]

__all__ = {x for x in ["Hello", "world"]}  # [invalid-all-format]

__all__ = foo  # [invalid-all-format]

__all__ = foo.bar  # [invalid-all-format]

__all__ = foo["bar"]  # [invalid-all-format]

__all__ = (foo := bar)  # [invalid-all-format]

__all__ = ["Hello"]

__all__ = ("Hello",)

__all__ = ["Hello"] + ("world",)

__all__ = [x for x in ["Hello", "world"]]

__all__ = list(["Hello", "world"])

__all__ = list({"Hello", "world"})

__all__ = list(["Hello"]) + list(["world"])

__all__ = tuple(["Hello"]) + ("world",)

__all__ = __all__ + ["Hello"]

__all__ = __all__ + multiprocessing.__all__

__all__ = list[str](["Hello", "world"])

__all__ = list[str](foo())

__all__ = (foo := ["Hello", "world"])
